### PR TITLE
Integrates stability reporting in github_reporting.py

### PR DESCRIPTION
**Pull request by deeplow**
- Adds an instability section when running `github_reporting.py` and `[unstable]` before failing but unstable tests in other sections (when `--instability` flag used)
- Creates a local DB of test data that will make queries easier to write

What's included:
1. Separate concerns in `github_reporting.py` into multiple files.
2. Convert Models into ORM database (with sqlalchemy) for making efficient local queries on historical data.
3. Separate child jobs from parent jobs (called orphan jobs) for better database modeling and querying
3. Updates `openqa_investigator.py` (instability plotting and reporting) to make use of the new code structure
4. Centralizes stability analysis code in `lib/instability_analysis.py`
5. Fixes minor issues in the plotting of stability data

Notes for reviewer:
  - when fetching a job it automatically fetches it's parents and children. This makes the script run slower. But as jobs get locally cached in the database this will progressively get faster
  - database migrations are not included. So whenever there is a db structure change one will need to delete the local `openqa_db.sqlite`. This may be annoying when checking the multiple commits individually as db repopulation will take quite a bit.
  - The stability analysis in `github_reporting.py` (with the `--instability` flag) will look for data in `updates` tests, which isn't currently populated. While it doesn't get populated the way to test it is by changing `updates` to `pull requests` in the `instability_analysis.py`'s line of `jobs = OpenQA.get_n_jobs_like(job, n=5, , flavor_override="update")`.

### CODE REVIEW

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
openqa_investigator seems to be broken:
```
python3 ./openqa_investigator.py --suite system_tests_network --test '.*/.*proxyvm_nm' --last 15   --output plot_templates
```
Figure_1

File: utils/lib/openqa_api.py [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+                    test_suite, max_history_len, version, flavor)
+        if not job_ids:
+            return []
+
+        if reference_job.job_id in job_ids:
+            ref_job_index = job_ids.index(reference_job.job_id)
+        else:
+            # try to find job with closest id (indicating temporal closeness)
+            ref_job_distance = map(
+                lambda job: reference_job.job_id - job.job_id,
```
[the remark this answers is not in the document]
Comment (deeplow) [reply utils/lib/openqa_api.py:780]: > This last commit doesn't work. job is just int here, not a job object.

Whoops. Should be fixed in commit 7829f48

> Furthermore, I think you should look for the closest job before reference_job, not just the closest one. Currently in practice it looks for a most recent job (the one with most negative difference here).

Fixed now in f193e04

File: utils/lib/openqa_api.py [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+                job_ids)
+            min_ref_job_dist = min(ref_job_distance)
+            ref_job_index = job_ids.index(min_ref_job_dist)
```
[the remark this answers is not in the document]
Comment (deeplow) [reply utils/lib/openqa_api.py]: Should be fixed now

**Comment by deeplow** [comment]
Thanks for the review. All feedback has now been addressed.

> openqa_investigator seems to be broken:
> 
> ```
> python3 ./openqa_investigator.py --suite system_tests_network --test '.*/.*proxyvm_nm' --last 15   --output plot_templates
> ```

Should be fixed now in 214866f44d6ef315661510c31f0380689b7a7c0f

File: utils/lib/openqa_api.py [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+        return relevant_jobs
+
+
+def config_db_session(in_memory=True, debug_db=False):
+
+    if in_memory:
+        db_engine = create_engine("sqlite:///:memory:", echo=debug_db)
+        Base.metadata.create_all(db_engine)
+    else:
+        db_file = "openqa_db.sqlite"
```
Comment (marmarek) [inline utils/lib/openqa_api.py:806]: Can you make the location configurable? Maybe env variable?

Comment (deeplow) [reply utils/lib/openqa_api.py]: Done. Did it both as an env var and as a parameter

File: utils/github_reporting.py [hunk 4]
```diff
@@ -742,9 +156,29 @@ def main():
...
+
+    parser.add_argument(
+        '--verbose',
+        action='store_true',
+        help="Enable debug logging."
+    )
+
+    parser.add_argument(
+        '--db-path',
+        default=os.getenv("LOCAL_OPENQA_CACHE_PATH"),
```
Comment (marmarek) [inline utils/github_reporting.py:177]: Missing `os` import...

Comment (deeplow) [reply utils/github_reporting.py:178]: Whoops

**Review by marmarek — APPROVED** [review]

**Pull request merged**